### PR TITLE
Return supported file type dictionary from CDOGS and throw 422 for unsupported conversion attempts

### PR DESCRIPTION
**Pull request by TimCsaky**
# Supported file types API endpoint

- Create new endpoint to return dictionary of supported file type conversions (https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-608)

- Updated API spec (https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-605)

- Added Unsupported conversion 422 response (https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-607)

## Types of changes
Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have checked that unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the OpenAPI 3.0 `v*.api-spec.yaml` documentation (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

## Further comments

See separate PR in nr-get-token repo for 'Update the postman collection in GETOK' (https://apps.nrs.gov.bc.ca/int/jira/browse/SHOWCASE-606)

### CODE REVIEW

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-idcqvl-tools.pathfinder.gov.bc.ca/dashboard?id=common-document-generation-service-pr-21

File: app/src/docs/v1.api-spec.yaml [hunk 1]
```diff
@@ -52,6 +52,31 @@ paths:
...
+    get:
+      summary: Returns a dictionary of supported input template file types and output file types.
+      description: >-
+        This endpoint checks the supported file types defined by the Carbone JS library that CDOGS depends on.
+      operationId: getFileTypes
+      tags:
+        - FileTypes
+      responses:
+        "200":
+          description: Returns a multi-dimensional array of the supported input Template file types and corresponding output file types available.
```
Comment (jujaga) [inline app/src/docs/v1.api-spec.yaml]: Description should be more succinct - I'd suggest
```
Returns the supported combinations of input templates and output file types
```

File: app/src/docs/v1.api-spec.yaml [hunk 2]
```diff
@@ -191,6 +216,22 @@ components:
...
+            type: string
+          outputFileTypes:
+            type: array
+            items:
+              type: string
+      example:
+       xlsx:
+         [xslx, docx, pdf]
+       docx:
+         [docx, pdf]
```
Comment (jujaga) [inline app/src/docs/v1.api-spec.yaml]: Missing a space in indentation
Example values should be in alphabetical order

File: app/src/components/fileTypes.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+const log = require('npmlog');
```
Comment (loneil) [inline app/src/components/fileTypes.js]: Doens't look like log is used in this script? Does lint catch the unused variable?

Comment (jujaga) [reply app/src/components/fileTypes.js]: Lint should be catching it - and if it isn't, I know vscode should be highlighting it if you have the eslint plugin installed.

File: app/src/components/fileTypes.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+const fileTypesDictionary = {
+  "csv":["doc", "docx", "html", "odt", "pdf", "rtf", "txt", "csv"],
+  "docx":["doc", "docx", "html", "odt", "pdf", "rtf", "txt"],
+  "html":["html", "odt", "pdf", "rtf", "txt"],
+  "odt":["doc", "docx", "html", "odt", "pdf", "rtf", "txt"],
+  "pptx":["odt", "pdf"],
+  "rtf":["docx","pdf"],
+  "txt":["doc", "docx", "html", "odt", "pdf", "rtf", "txt"],
+  "xlsx":["odt", "pdf", "rtf", "txt", "csv", "xls", "xlsx"]
+};
```
Comment (jujaga) [inline app/src/components/fileTypes.js]: Style: there needs to be a space between the colon and the open brace. Assuming your IDE is set up correctly it should be able to auto-format that for you.

File: app/src/routes/v1.js [hunk 5]
```diff
@@ -14,6 +15,7 @@ router.get('/', (_req, res) => {
   res.status(200).json({
     endpoints: [
       '/health',
+      '/fileTypes',
```
Comment (jujaga) [inline app/src/routes/v1.js]: Style: alphabetize this list

File: app/src/routes/v1.js [hunk 6]
```diff
@@ -33,6 +35,9 @@ router.get('/api-spec.yaml', (_req, res) => {
 /** Health Router */
 router.use('/health', keycloak.protect(), healthRouter);
 
+/** File Types Router */
+router.use('/fileTypes', keycloak.protect(), fileTypesRouter);
+
```
Comment (jujaga) [inline app/src/routes/v1.js]: Style: I suggest this be above the health router for alphabetical ordering reasons.

File: app/src/components/validators.js [hunk 7]
```diff
@@ -82,7 +83,22 @@ const models = {
...
+      const fileTypes = fileTypesComponent.fileTypesDictionary;
+      // if input/output combination not in dictionary then error
+      if(!fileTypes[inputFileType].includes(outputFileType) ){    
+        log.error('File conversion not supported');
+        return false;
+      }
+      else{
+        return true;
+      }
+    }
```
Comment (jujaga) [inline app/src/components/validators.js]: This would be equivalent and be able to avoid multiple return statements
``` js
    fileConversion: (inputFileType, outputFileType) => {  
      // get dictionary
      const fileTypes = fileTypesComponent.fileTypesDictionary;
      // if input/output combination not in dictionary then error
      const isSupported = fileTypes[inputFileType].includes(outputFileType);
      if (isSupported) {
        log.error('File conversion not supported');
      }
      return isSupported;
    } 
```

File: app/src/routes/v1/fileTypes.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
+const fileTypesRouter = require('express').Router();
+const fileTypesComponent = require('../../components/fileTypes');
+
+/** Returns the dictionary of input/output file types */
+fileTypesRouter.get('/', async (_req, res, next) => {
+  
+  const fileTypes = await fileTypesComponent.getFileTypes();
```
Comment (jujaga) [inline app/src/routes/v1/fileTypes.js]: Based on what we know about that component, we don't need to do the async/await wrapper on this as that component (or soon to be JSON file resource) can be synchronous call. Can remove the async declaration on the callback function declaration on line 5 as well as there's no awaits in this function at all.

File: app/tests/unit/components/validators/validators.customValidators.spec.js [hunk 9]
```diff
@@ -59,4 +60,28 @@ describe('customValidators.docGen', () => {
...
+        outputFileType: 'ppt',
+      }
+    };
+
+    const result = await customValidators.docGen(body);
+
+    expect(result).toBeTruthy();
+    expect(Array.isArray(result)).toBeTruthy();
+    expect(result.length).toEqual(1);
+    expect(result[0].message).toMatch('Unsupported');
```
Comment (jujaga) [inline app/tests/unit/components/validators/validators.customValidators.spec.js]: It's probably better to match on a more complete string than just a part of it as it'd be more "correct"
`Unsupported file type conversion. A dictionary of supported input and output file types can be found at API endpoint \'/fileTypes\'`. Even matching on the first sentence would be better as we want to ensure the right message is being returned in the array - there could be cases in the future where there's multiple unsupported messages, and we wouldn't know which type of unsupported it was.

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-idcqvl-tools.pathfinder.gov.bc.ca/dashboard?id=common-document-generation-service-pr-21

File: app/tests/unit/components/validators/validators.customValidators.spec.js [hunk 10]
```diff
@@ -51,6 +51,7 @@ describe('customValidators.docGen', () => {
       template: {
         content: 'ZHNmc2Rmc2RmZHNmc2Rmc2Rmc2Rm',
         contentFileType: 'docx',
+        outputFileType: 'docx',
```
Comment (loneil) [inline app/tests/unit/components/validators/validators.customValidators.spec.js]: Was outputFileType added to get this existing test to pass? Because outputFileType is not a required field, so this test is meant to pass without any optional fields. If the test was failing then I think it's probably evidence that the validator is not behaving as intended. Is the validator possibly requiring an outputFileType even though it's an optional field?

Comment (TimCsaky) [reply app/tests/unit/components/validators/validators.customValidators.spec.js]: yes. i added outputFileType as a way to get that test to pass. 
I'll take a look a fixing the validator now

**Comment by TimCsaky** [comment]
this pr has been a very useful learning experience.

File: app/src/routes/v1/fileTypes.js [hunk 11]
```diff
@@ -0,0 +1,16 @@
+const fileTypesRouter = require('express').Router();
+const fileTypesObject = require('../../assets/fileTypes.json');
+
+/** Returns the dictionary of input/output file types */
+fileTypesRouter.get('/', async (_req, res, next) => {
+
```
Comment (jujaga) [inline app/src/routes/v1/fileTypes.js]: Unnecessary whitespace line

File: app/src/routes/v1/fileTypes.js [hunk 12]
```diff
@@ -0,0 +1,16 @@
+const fileTypesRouter = require('express').Router();
+const fileTypesObject = require('../../assets/fileTypes.json');
+
+/** Returns the dictionary of input/output file types */
+fileTypesRouter.get('/', async (_req, res, next) => {
```
Comment (jujaga) [inline app/src/routes/v1/fileTypes.js]: This doesn't need to be an asynchronous function.
```
fileTypesRouter.get('/', (_req, res, next) => {
```

File: app/src/components/validators.js [hunk 13]
```diff
@@ -32,9 +33,9 @@ const models = {
       return true;
     },
 
-    /** @function contentFileType is required */
+    /** @function contentFileType is required and exists as a valid input file type defined in the fileTypes Dictionary */
     contentFileType: value => {
-      return validatorUtils.isNonEmptyString(value);
+      return validatorUtils.isNonEmptyString(value) && value in fileTypes;
```
Comment (loneil) [inline app/src/components/validators.js]: Consider adding a unit test specifically for the "value not in fileTypes" failure condition in the `describe('models.template.contentFileType` suite in validators.models.spec.js

Comment (TimCsaky) [reply app/src/components/validators.js]: Thank you dude.. they were good suggestions. i believe I've corrected that now

**Comment by TimCsaky** [comment]
> 
> 
> Just remembered about case-sensitivity. Looks like if the consumer supplies "DOCX" as a contentfiletype it's rejected by the validator. So I think we'd want case-insensitive compares for file extensions on content and output file type.
> If someone consuming our API has a user supply a file named "FILE.DOCX" and parses the extension as-is then I think we'd want to accommodate the upper case.

good thought. i've updated the validator

**Comment by bcgov-nr-csst** [comment]
SonarQube Analysis: https://sonarqube-idcqvl-tools.pathfinder.gov.bc.ca/dashboard?id=common-document-generation-service-pr-21
